pkg/state/eth_info.go

```diff
@@ -0,0 +1,107 @@
+package state
+
+import (
+	"github.com/pkg/errors"
+	"github.com/wavesplatform/gowaves/pkg/errs"
+	"github.com/wavesplatform/gowaves/pkg/proto"
+	"github.com/wavesplatform/gowaves/pkg/proto/ethabi"
+	"github.com/wavesplatform/gowaves/pkg/settings"
+)
+
+const (
+	EthereumTransferWavesKind = iota + 1
+	EthereumTransferAssetsKind
+	EthereumInvokeKind
+)
+
+type ethInfo struct {
+	stor     *blockchainEntitiesStorage
+	settings *settings.BlockchainSettings
+}
+
+func newEthInfo(stor *blockchainEntitiesStorage, settings *settings.BlockchainSettings) *ethInfo {
+	return &ethInfo{stor: stor, settings: settings}
+}
+
+func GuessEthereumTransactionKind(
+	data []byte,
+	to *proto.EthereumAddress,
+	newestAssetInfo func(id proto.AssetID, filter bool) (*assetInfo, error)) (int64, error) {
+	if data == nil {
+		return EthereumTransferWavesKind, nil
+	}
+
+	selectorBytes := data
+	if len(data) < ethabi.SelectorSize {
+		return 0, errors.Errorf("length of data from ethereum transaction is less than %d", ethabi.SelectorSize)
+	}
+	selector, err := ethabi.NewSelectorFromBytes(selectorBytes[:ethabi.SelectorSize])
+	if err != nil {
+		return 0, errors.Errorf("failed to guess ethereum transaction kind, %v", err)
+	}
+
+	assetID := (*proto.AssetID)(to)
+
+	_, err = newestAssetInfo(*assetID, true)
+	if err != nil && !errors.Is(err, errs.UnknownAsset{}) {
+		return 0, errors.Errorf("failed to get asset info by ethereum recipient address %s, %v", to.String(), err)
+	}
+
+	if ethabi.IsERC20TransferSelector(selector) && err == nil {
+		return EthereumTransferAssetsKind, nil
+	}
+
+	return EthereumInvokeKind, nil
+}
+
+func (e *ethInfo) ethereumTransactionKind(ethTx *proto.EthereumTransaction, params *appendTxParams) (proto.EthereumTransactionKind, error) {
+	txKind, err := GuessEthereumTransactionKind(ethTx.Data(), ethTx.To(), e.stor.assets.newestAssetInfo)
+	if err != nil {
+		return nil, errors.Errorf("failed to guess ethereum tx kind, %v", err)
+	}
+
+	switch txKind {
+	case EthereumTransferWavesKind:
+		return proto.NewEthereumTransferWavesTxKind(), nil
+	case EthereumTransferAssetsKind:
+
+		db := ethabi.NewErc20MethodsMap()
+		decodedData, err := db.ParseCallDataRide(ethTx.Data())
+		if err != nil {
+			return nil, errors.Errorf("failed to parse ethereum data")
+		}
+		if len(decodedData.Inputs) != ethabi.NumberOfERC20TransferArguments {
+			return nil, errors.Errorf("the number of arguments of erc20 function is %d, but expected it to be %d", len(decodedData.Inputs), ethabi.NumberOfERC20TransferArguments)
+		}
```

Comment:
You can use `ethabi.GetERC20TransferArguments` and store result into `EthereumTransferAssetsErc20TxKind` structure